nep-17.mediawiki

```diff
@@ -0,0 +1,107 @@
+
+==Specification==
+
+In the method definitions below, we provide both the definitions of the functions as they are defined in the contract as well as the invoke parameters.
+
+===Methods===
+
+====symbol====
+
+<pre>
+public static string symbol()
+</pre>
+
+Returns a short string symbol of the token managed in this contract. e.g. <code>"MYT"</code>. This symbol SHOULD be short (3-8 characters is recommended), with no whitespace characters or new-lines and SHOULD be limited to the uppercase latin alphabet (i.e. the 26 letters used in English).
+
+This method MUST always return the same value every time it is invoked.
+
+====decimals====
+
+<pre>
+public static byte decimals()
+</pre>
+
+Returns the number of decimals used by the token - e.g. <code>8</code>, means to divide the token amount by <code>100,000,000</code> to get its user representation.
+
+This method MUST always return the same value every time it is invoked.
+
+====totalSupply====
+
+<pre>
+public static BigInteger totalSupply()
+</pre>
+
+Returns the total token supply deployed in the system.
+
+====balanceOf====
+
+<pre>
+public static BigInteger balanceOf(byte[] account)
+</pre>
+
+Returns the token balance of the <code>account</code>.
+
+The parameter <code>account</code> MUST be a 20-byte address. If not, this method SHOULD <code>throw</code> an exception.
+
+If the <code>account</code> is an unused address, this method MUST return <code>0</code>.
+
+====transfer====
+
+<pre>
+public static bool transfer(byte[] from, byte[] to, BigInteger amount)
+</pre>
+
+Transfers an <code>amount</code> of tokens from the <code>from</code> account to the <code>to</code> account.
+
+The parameters <code>from</code> and <code>to</code> MUST be 20-byte addresses. If not, this method SHOULD <code>throw</code> an exception.
+
+The parameter <code>amount</code> MUST be greater than or equal to <code>0</code>. If not, this method SHOULD <code>throw</code> an exception.
+
+The function MUST return <code>false</code> if the <code>from</code> account balance does not have enough tokens to spend.
+
+If the method succeeds, it MUST fire the <code>transfer</code> event, and MUST return <code>true</code>, even if the <code>amount</code> is <code>0</code>, or <code>from</code> and <code>to</code> are the same address.
+
+The function SHOULD check whether the <code>from</code> address equals the caller contract hash. If so, the transfer SHOULD be processed; If not, the function SHOULD use the SYSCALL <code>Neo.Runtime.CheckWitness</code> to verify the transfer.
+
+If the transfer is not processed, the function MUST return <code>false</code>.
+
+The function MUST call <code>onPayment</code> method AFTER fire the <code>Transfer</code> event if the receiver is a deployed contract. If the receiver doesn't want to receive this transfer it MUST call <code>ABORT</code>.
```

Comment:
I think we can leave receiver behavior (and `onPayment` definition) for another NEP (because the receiver doesn't have to support full NEP-17), here it's important to call `onPayment`, but it's not clear from this document how a contract should detect `if the receiver is a deployed contract`.

Reply:
We only need to explain the logic clearly in the proposal. As for how to implement it, there can be templates or examples. The specific way is to call `System.Blockchain.GetContract`, and then check the return value.

Reply:
`System.Blockchain.GetContract` seems to be too heavy for the task to me (because we don't really need contract's data), but maybe it's OK (0.01 GAS, same as notification, and it's not hard to `DROP` the result). And you're right, if we're resorting to these standard APIs they can be omitted from the specification.

What do you think about making an additional companion NEP (NEP-18, for example) specifically for `onPayment`? It can simplify things in that this NEP will refer to it for `onPayment` behavior definition and contracts could then easily declare that they're supporting NEP-18 (meaning that they have proper `onPayment` and can process incoming tokens). For example, NEO and GAS contracts are NEP-17-compatible, but not NEP-18 (they don't have `onPayment`), while NeoFS mainnet contract is the opposite, it's not NEP-17 (it doesn't declare any token of its own), but it does support NEP-18 (can accept payments to it).

Reply:
I think it is a good idea, @roman-khimov.

Reply:
> What do you think about making an additional companion NEP (NEP-18, for example) specifically for onPayment?

I don't think it's necessary. We directly check whether it contains the `onPayment` method, without checking whether it implements a certain NEP.